python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -707,18 +708,56 @@ def _submit_runs_and_update_backfill_in_chunks(
-        for run_request in run_requests_chunk:
+        run_requests_i = 0
+        while run_requests_i < len(run_requests_chunk):
+            run_request = run_requests_chunk[run_requests_i]
             yield None
-            submit_run_request(
-                run_request=run_request,
-                asset_graph=asset_graph,
-                # create a new request context for each run in case the code location server
-                # is swapped out in the middle of the backfill
-                workspace=workspace_process_context.create_request_context(),
-                instance=instance,
-                pipeline_and_execution_plan_cache=pipeline_and_execution_plan_cache,
+
+            # create a new request context for each run in case the code location server
+            # is swapped out in the middle of the backfill
+            workspace = workspace_process_context.create_request_context()
+            execution_data = get_job_execution_data_from_run_request(
+                asset_graph,
+                run_request,
+                instance,
+                workspace=workspace,
+                run_request_execution_data_cache=run_request_execution_data_cache,
             )
 
+            if _execution_plan_targets_asset_selection(
+                execution_data.external_execution_plan.execution_plan_snapshot,
+                check.not_none(run_request.asset_selection),
+            ):
+                submit_run_request(
+                    run_request=run_request,
+                    workspace=workspace,
+                    instance=instance,
+                    run_request_execution_data=execution_data,
+                )
+                run_requests_i += 1
+
+            elif num_retries_allowed > 0:
+                logger.warning(
+                    "Execution plan is out of sync with the workspace. Pausing the backfill for "
+                    "60s to allow the execution plan to rebuild with the updated workspace."
+                )
+                # Sleep for RELOAD_WORKSPACE_INTERVAL seconds since the workspace can be refreshed
+                # at most once every interval
+                time.sleep(RELOAD_WORKSPACE_INTERVAL)
+                # Clear the execution plan cache as this data is no longer valid
+                run_request_execution_data_cache = {}
+                num_retries_allowed -= 1
+                # If the execution plan does not targets the asset selection, the asset graph
+                # likely is outdated and targeting the wrong job, refetch the asset
+                # graph from the workspace
+                workspace = workspace_process_context.create_request_context()
+                asset_graph = ExternalAssetGraph.from_workspace(workspace)
+
+            else:  # Already hit the max number of retries
+                check.failed(
+                    f"Failed to target asset selection {run_request.asset_selection} in run after {num_retries_allowed} retries."
```

Comment:
nit: this will always say "after 0 retries" because `num_retries_allowed` is the value being decremented.

Reply:
nice catch